src/base_emitter.iced

```diff
@@ -30,6 +31,71 @@ exports.BaseEmitter = class BaseEmitter
     else
       @_code.push("") if (line is "}" or line is ")") and @_tabs is 0
 
+
+  get_field_type : (type) ->
+    if typeof type is 'string'
+      return type
+    else
+      if Array.isArray type
+        return type[1]
+      else if type.type is 'array'
+        return @get_field_type type.items
+      else if type.type is 'map'
+        return @get_field_type type.values
+
+
+  create_dep_graph : (types) ->
+    graph = {}
+    for type in types
+      graph[type.name] = {type, in_count: 0, children: []}
+
+    for type in types
+      switch type.type
+        when "record"
+          if type.typedef and not is_primitive type.typedef
+            graph[type.typedef].children = [type.name]
+            graph[type.name].in_count++
+          else
+            for field in type.fields
+              field_type = @get_field_type field.type
+              if not is_primitive(field_type) and not graph[field_type].children.includes type.name
+                graph[field_type].children.push type.name
+                graph[type.name].in_count++
+        when "variant"
+          for case_ in type.cases
+            if not case_.body?
+              continue
+            case_type = if typeof case_.body == 'object' then case_.body.items else case_.body
+            if not is_primitive(case_type) and not graph[case_type].children.includes type.name
+              graph[case_type].children.push type.name
+              graph[type.name].in_count++
+
+    graph
+
+  # Python requires that types be declared before they can be used.
+  # This isn't a requirement in our avdl files, so we need to go through the types, create a dependency graph, and then sort it
+  # Other languages will get this for free; it doesn't affect them
+  sort_types : (types) ->
+    graph = @create_dep_graph types
+
+    # Topological sort
```

Comment:
I think this would be helpful to future readers
```
# Iteratively, release types that do not require any unreleased types. Until all types have been released.
```